### DISCUSSION TITLE
Insert List, Struct and other composite types with C Api

### DISCUSSION THREAD

**Discussion by Giorgi** [duckdb/duckdb#9365] [Ideas]
With [Prepared Statements](https://duckdb.org/docs/api/c/prepared) it's possible to insert primitive values but there are no functions to bind list, struct, union, map values and insert them into a database.

We need functions to insert these data types as well as nested variants (list of list, list of struct, struct with lists, etc)

**Comment by karlseguin** [comment]
https://github.com/duckdb/duckdb/discussions/7482 has this request specifically for lists. This more general discussion is probably better though.

**Comment by karlseguin** [comment]
It would also be nice to get all of this in the Appender APIs, which currently have the same limitations as Prepared Statements.

**Comment by Giorgi** [comment]
@Mause Since there is already `duckdb_append_data_chunk` that allows us to write "raw" data to an appender, something similar can be done here for prepared statements. The major will be that in the case of `duckdb_append_data_chunk` the data chunk can have multiple rows (size>=1) while for prepared statements the chunk will have size = 1.

The API flow would look like this: Once we create a prepared statement, we call `duckdb_param_logical_type` (new function) to get the logical types for parameters. Once we have the types, we can use the existing `duckdb_create_data_chunk` function to create the chunk and write to the chunk as we already do in other cases. Finally, we call `duckdb_bind_data_chunk` (new function) to bind the chunk to a prepared statement.

In case of auto-incremented (?) or positional ($1) parameters, the index for `duckdb_param_logical_type` and `duckdb_data_chunk_get_vector` will be the index of the parameter. In the case of named parameters, we will first call `duckdb_bind_parameter_index` to get the index of the parameter.

Thoughts?

**Reply by Tishj** [reply]
`duckdb_param_type` already exists

I think all of this already exists, apart from `duckdb_bind_data_chunk`
It feels a little disconnected from the way prepared parameters work currently, as they are a ConstantExpression, containing a Value.

But I do like how using a DataChunk for this allows it to be easier extended eventually to support binding multiple rows, which lines up a bit with <https://github.com/duckdb/duckdb/discussions/11755>

**Reply by Giorgi** [reply]
> `duckdb_param_type` already exists

That one doesn't return a logical type. Logical type is needed to create a data chunk.

**Reply by Tishj** [reply]
Hmm I see, this just returns an Enum pretty much, which is part of the LogicalType (LogicalTypeId) but doesn't paint the full picture

**Reply by Giorgi** [reply]
Yes, we need just two new functions to enable composite types in prepared statements. `duckdb_bind_data_chunk` may feel a bit disconnected but it's the same approach as used in the appender.

**Reply by karlseguin** [reply]
I'm a fan of any additions which allows binding more complex types to a prepared statement. But, to play the devils advocate...

Writing data chunks is considerably more complicated. We can currently bind a 36-character hex-encoded UUID using the `duckdb_bind_varchar` function, but doing so with vectors requires re-implementing the uuid -> i128 logic in each driver. This is hardly the only or most complicated conversion.

Even for those implementations that have an Appender API already using the data chunk and thus can re-use that logic, I don't think we're close to the same ease-of-use / flexibility that application developers would expect. 

Unless I misunderstand, I'm afraid this proposal would result in 1 API that offers completeness and another that offers simplicity and consistency (consistent between implementations, and between what application developers and end-users expect from an SQL driver).

**Reply by Giorgi** [reply]
@karlseguin Those are valid concerns, it would be great if we could use `duckdb_bind_xxx` for some parameters and data chunk for others. Not sure if it is possible but an API like this would be great:

`duckdb_bind_data_chunk(duckdb_prepared_statement prepared_statement, idx_t[] param_indexes, duckdb_data_chunk chunk);`

The indexes array would indicate which parameters should be populated from the data chunk.

This way we would use data chunk only for composite types and the "easy" `duckdb_bind_xxx` functions for other parameters.

What do you think about this approach @Mytherin ?

**Reply by Mytherin** [reply]
In the C++ API we use the `Value` type for prepared statements - similarly the C API has the `duckdb_value` type. My current design philosophy is to have the C API match the C++ API as much as possible - as such I think the correct path forward is to extend the `duckdb_value` type.

**Reply by Giorgi** [reply]
So you mean extending the [Value Interface](https://duckdb.org/docs/api/c/api#value-interface) so that we can create arbitrarily nested composite values?

**Reply by Mytherin** [reply]
Exactly, that is actually already possible with `duckdb_create_struct_value`, `duckdb_create_list_value`, `duckdb_create_array_value`. It seems this API is mostly missing primitive types.

**Reply by Mytherin** [reply]
We could also consider adding a `duckdb_value duckdb_vector_get_value(duckdb_vector, idx_t row)` method that allows you to get a value from a vector. We have that in C++ as well.
